File: src/clu/sources/proc_cpuinfo.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_file

log = logging.getLogger(__name__)


class ProcCpuinfo(Source):
    @classmethod
    def __has_flags(cls, check_flags, all_flags):
        check_flags = check_flags.split()
        all_flags = all_flags.split()
        log.debug(f"count is {len(check_flags)} {len(all_flags)}")
        log.debug(f"{check_flags=}")
        log.debug(f"{all_flags=}")
        for flag in check_flags:
            if flag not in all_flags:
                return False
        return True

    def provides(self, provides: Provides) -> None:
        provides["phy.cpu.arch_version"] = self

    def requires(self, requires: Requires) -> None:
        requires.files.append("/proc/cpuinfo")

    def parse(self, facts: Facts) -> None:
        if facts["phy.arch"] not in ("x86_64", "amd64"):
            log.info("Not an x86_64/amd64 architecture, skipping cpuinfo flags parsing")
            return

        vers = [
            "lm cmov cx8 fpu fxsr mmx syscall sse2",
            "cx16 lahf_lm popcnt sse4_1 sse4_2 ssse3",
            "avx avx2 bmi1 bmi2 f16c fma abm movbe xsave",
            "avx512f avx512bw avx512cd avx512dq avx512vl",
        ]

        data = text_file("/proc/cpuinfo")
        # log.debug(f"data={data.splitlines() if data else []}") This can be HUGE
        cpu_flags = ""
        if data:
            for line in data.splitlines():
                if line.startswith("flags"):
                    cpu_flags = line.split(":", 1)[1].strip()
                    break
        log.debug(f"cpu_flags={cpu_flags.split()}")
        cpu_version = 0
        for idx, v in enumerate(vers):
            if self.__has_flags(v, cpu_flags):
                cpu_version += 1
            else:
                break
        facts["phy.cpu.arch_version"] = f"x86_64_v{cpu_version}"

File: pkg/sources/proc_cpuinfo.go
package sources

// Go port of src/clu/sources/proc_cpuinfo.py
// Derives an x86_64 architecture micro-version (x86_64_vN) based on presence of
// progressively demanding CPU flags in /proc/cpuinfo. Skips non-x86_64 architectures.

import (
	"strconv"
	"strings"

	pkg "github.com/huntermatthews/clu/pkg"
)

// ProcCpuinfo determines phy.cpu.arch_version for x86_64/amd64 platforms.
type ProcCpuinfo struct{}

// Provides registers the architecture version fact.
func (p *ProcCpuinfo) Provides(pr pkg.Provides) { pr["phy.cpu.arch_version"] = p }

// Requires declares /proc/cpuinfo file dependency.
func (p *ProcCpuinfo) Requires(r *pkg.Requires) { r.Files = append(r.Files, "/proc/cpuinfo") }

// hasAllFlags checks whether every flag in required (space separated) exists in allFlags.
func hasAllFlags(required, allFlags string) bool {
	req := strings.Fields(required)
	all := strings.Fields(allFlags)
	// Build a set for quick membership tests.
	set := make(map[string]struct{}, len(all))
	for _, f := range all {
		set[f] = struct{}{}
	}
	for _, f := range req {
		if _, ok := set[f]; !ok {
			return false
		}
	}
	return true
}

// Parse implements the flag progression logic.
func (p *ProcCpuinfo) Parse(f *pkg.Facts) {
	arch, _ := f.Get("phy.arch")
	if arch != "x86_64" && arch != "amd64" {
		return // skip non-x86
	}

	versions := []string{
		"lm cmov cx8 fpu fxsr mmx syscall sse2",
		"cx16 lahf_lm popcnt sse4_1 sse4_2 ssse3",
		"avx avx2 bmi1 bmi2 f16c fma abm movbe xsave",
		"avx512f avx512bw avx512cd avx512dq avx512vl",
	}

	data, err := pkg.FileReader("/proc/cpuinfo")
	flagsField := ""
	if err == nil && data != "" {
		for _, line := range strings.Split(data, "\n") {
			if strings.HasPrefix(line, "flags") {
				parts := strings.SplitN(line, ":", 2)
				if len(parts) == 2 {
					flagsField = strings.TrimSpace(parts[1])
				}
				break
			}
		}
	}
	version := 0
	for _, group := range versions {
		if hasAllFlags(group, flagsField) {
			version++
		} else {
			break
		}
	}
	f.Set("phy.cpu.arch_version", "x86_64_v"+strconv.Itoa(version))
}
